FHEM/14_SD_BELL.pm

```diff
@@ -0,0 +1,303 @@
+		readingsBulkUpdateIfChanged
+		readingsDelete
+		readingsEndUpdate
+		readingsSingleUpdate
+		ReadingsVal
+		Value
+    ))
+};
+
+
+
+#############################
+sub Define($$) {
+	my ($hash, $def) = @_;
+	my @a = split("[ \t][ \t]*", $def);
+
+	# Argument					    0	   1		2		    3				4
+	return "wrong syntax: define <name> SD_BELL <Protocol> <HEX-Value> <optional IODEV>" if(int(@a) < 3 || int(@a) > 5);
+	return "wrong <protocol> $a[2]" if not($a[2] =~ /^(?:14|15|32|41|57|79)/s);
+	### checks ###
+	return "wrong HEX-Value! Protocol $a[2] HEX-Value to short | long or not HEX (0-9 | a-f | A-F){3,8}" if (not $a[3] =~ /^[0-9a-fA-F]{3,8}/s);
+
+	
+	$hash->{lastMSG} =  "";
+	$hash->{bitMSG} =  "";
+	my $iodevice = $a[4] if($a[4]);
+	my $name = $hash->{NAME};
+
+	$modules{SD_BELL}{defptr}{$hash->{DEF}} = $hash;
+	my $ioname = $modules{SD_BELL}{defptr}{ioname} if (exists $modules{SD_BELL}{defptr}{ioname} && not $iodevice);
+	$iodevice = $ioname if not $iodevice;
+	
+	### Attributes | model set after codesyntax ###
+	my $devicetyp = $a[2];
+	$devicetyp = $models{$a[2]};
+	$attr{$name}{model}	= $devicetyp if( not defined( $attr{$name}{model} ) );	
+	$attr{$name}{room}	= "SD_BELL"	if( not defined( $attr{$name}{room} ) );
+	
+	AssignIoPort($hash, $iodevice);
+}
+
+###################################
+sub Set($$$@) {
+	my ( $hash, $name, @a ) = @_;
+	my $cmd = $a[0];
+	my $ioname = $hash->{IODev}{NAME};
+	my $model = AttrVal($name, "model", "unknown");
+	my @split = split(" ", $hash->{DEF});
+	my $protocol = $split[0];
+	my $ret = undef;
+	
+	if ($hash->{bitMSG} ne "") {
+		if ($cmd eq "?") {
+			$ret .= "ring:noArg";
+		} else {
+			my $msg = "P$protocol#" . $hash->{bitMSG};
+			$msg .= "#R5";		# Anzahl Wiederholungen noch klären!
+			Log3 $name, 4, "$ioname: $name sendMsg=$msg";
+
+			if ($cmd ne "?") {
+				$cmd = "ring";
+			}				
+
+			Log3 $name, 3, "$ioname: $name set $cmd" if ($cmd ne "?");
+			IOWrite($hash, 'sendMsg', $msg);
+		}
+	} else {
+		return $ret;
+	}
+
+	readingsSingleUpdate($hash, "state" , $cmd, 1) if ($cmd ne "?");
+	return $ret;
+}
+
+#####################################
+sub Undef($$) {
+	my ($hash, $name) = @_;
+	delete($modules{SD_BELL}{defptr}{$hash->{DEF}})
+		if(defined($hash->{DEF}) && defined($modules{SD_BELL}{defptr}{$hash->{DEF}}));
+	return undef;
+}
+
+
+###################################
+sub Parse($$) {
+	my ($iohash, $msg) = @_;
+	my $ioname = $iohash->{NAME};
+	my ($protocol,$rawData) = split("#",$msg);
+	$protocol=~ s/^[u|U|P](\d+)/$1/;		# extract protocol ID, $1 = ID
+	my $hlen = length($rawData);
+	my $blen = $hlen * 4;
+	my $bitData = unpack("B$blen", pack("H$hlen", $rawData));
+	
+	Log3 $iohash, 4, "$ioname: SD_BELL protocol $protocol $models{$protocol}, bitData $bitData";
+	
+	my $def;
+	my $deviceCode = $rawData;
+	my $devicedef;
+	my $state;
+	
+	if (!$def) {
+		$devicedef = $protocol . " " .$deviceCode;
+		$def = $modules{SD_BELL}{defptr}{$devicedef};
+	}
+	
+	$modules{SD_BELL}{defptr}{ioname} = $ioname;
```

Comment:
Wäre das identisch mit $iohash->{NAME} ?

Reply:
Da bin ich mir jetzt nicht sicher, ob wir das so machen sollten.

Zitat aus dem FHEM Development Wiki:

> Dazu wird in der Regel im Modul-Hash im modulspezifischen Berreich eine Liste defptr (Definition Pointer) geführt, welche jede eindeutige Adresse/ID dem entsprechenden Geräte-Hash zuordnet

Wenn ich es jetzt richtig verstehe, dann verweisen wir nicht auf einen Hash sondern speicher den Namen des iohash ab. Ich hab noch nicht geblickt, wieso wir den im defptr ablegen und nicht im hash der Geräteinstanz.

In dieser Form würde es vermutlich auch zu seiteneffekten bei mehr als einem iodev kommen oder?

Reply:
Der Grund wieso ich das so gelöst habe und auch schon an anderen Stellen einsetzte ist folgender.
Sobald ich an meinem System mehrere Empfänger dran hängen habe, so würde mit bei einem Anlegen des Devices immer ein "beliebiger" Empfänger ins IODev geschrieben werden welcher zur Zeit dran steckt. Ich möchte dort im IODEV aber den Empfänger sofort drin stehen haben worüber ich das Device emfpangen und angelegt bekommen habe.

Sieh diesen defpr nur als "Merker" um auch das richtige Device sofort ins IODev attr zu schreiben.